File: python/cursor.py
import constants
import math
from table import Table
from node import Node
from leaf_node import Leaf_Node
from internal_node import Internal_Node
from row import Row

class Cursor:

  def __init__(self, table, table_end = False):
    root_node = Node(table.get_root_page())
    num_root_node_cells = root_node.num_cells

    self.table = table
    self.page_num = self.table.root_page_num
    self.cell_num = 0
    self.end_of_table = num_root_node_cells == 0


  def get_page(self):
    return self.table.pager.get_page(self.page_num)

  def advance(self, num_node_cells):
    # let's just advance the cursor within the node for now. The table only has one node
    self.cell_num += 1
    self.end_of_table = (self.cell_num >= num_node_cells)


  def set_cell_num(self, node, key):
    if node.type is not Node.NODE_TYPE_LEAF:
      print("I still need to implement search through internal nodes")
      return

    self.cell_num = node.find_index(key)

  def find_page(self, key):
    node = Internal_Node(self.get_page())

    # If we've reached a leaf node, this is the page we're looking for
    if node.type == Node.NODE_TYPE_LEAF:
      return

    # If this isn't a leaf node, figure out where to find the next node in the tree
    self.page_num = node.get_child_page_num(key)

    self.find_page(key)

File: c/cursor.c
#include <fcntl.h>
#include <errno.h>
#include <unistd.h>
#include <stdbool.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#include "cursor.h"

Cursor* table_start(Table* table) {
  Cursor* cursor = malloc(sizeof(Cursor));
  cursor->table = table;
  cursor->page_num = table->root_page_num;
  cursor->cell_num = 0;

  void* root_node = get_page(table->pager, table->root_page_num);
  uint32_t num_cells = *leaf_node_num_cells(root_node);
  cursor->end_of_table = (num_cells == 0);

  return cursor;
}

/*
  Return the position of the given key.
  If the key is not present, return the position
  where it should be inserted
*/
Cursor* table_find(Table* table, uint32_t key) {
  uint32_t root_page_num = table->root_page_num;
  void* root_node = get_page(table->pager, root_page_num);

  if (get_node_type(root_node) == NODE_LEAF) {
    Cursor* cursor = malloc(sizeof(Cursor));
    cursor->table = table;
    cursor->page_num = root_page_num;
    cursor->cell_num = leaf_node_find_index(get_page(table->pager, root_page_num), key);
    return cursor;
  } else {
    printf("Need to implement searching an internal node\n");
    exit(EXIT_FAILURE);
  }
}

void* cursor_value(Cursor* cursor) {
  uint32_t page_num = cursor->page_num;
  void *page = get_page(cursor->table->pager, page_num);
  return leaf_node_value(page, cursor->cell_num);
}

void cursor_advance(Cursor* cursor) {
  void* node = get_page(cursor->table->pager, cursor->page_num);

  cursor->cell_num += 1;
  if (cursor->cell_num >= (*leaf_node_num_cells(node))) {
    cursor->end_of_table = true;
  }
}
